tests/Test/Xrefcheck/CanonicalPathSpec.hs

```diff
@@ -0,0 +1,62 @@
+{- SPDX-FileCopyrightText: 2022 Serokell <https://serokell.io>
+ -
+ - SPDX-License-Identifier: MPL-2.0
+ -}
+
+module Test.Xrefcheck.CanonicalPathSpec where
+
+import Universum
+
+import System.Directory (getCurrentDirectory)
+import Test.Tasty (TestTree, testGroup)
+import Test.Tasty.HUnit (testCase, (@?=))
+
+import Xrefcheck.System
+
+test_canonicalPath :: TestTree
+test_canonicalPath = testGroup "Path canonicalization"
+  [ testCase "Trailing separator" $ do
+      path <- canonicalizePath "./example/dir/"
+      current <- getCurrentDirectory >>= canonicalizePath
+      getRelativeOrAbsoluteChild current path @?= "example/dir",
+    testCase "Parent directory indirection" $ do
+      path <- canonicalizePath "dir1/../dir2"
+      current <- getCurrentDirectory >>= canonicalizePath
+      getRelativeOrAbsoluteChild current path @?= "dir2",
+    testCase "Through parent directory indirection" $ do
+      path <- canonicalizePath "dir1/../../../dir2"
+      current <- getCurrentDirectory >>= canonicalizePath
+      root <- current </ "../.."
+      getRelativeOrAbsoluteChild root path @?= "dir2",
+    testCase "Current directory indirection" $ do
+      path <- canonicalizePath "././dir1/./././dir2/././"
+      current <- getCurrentDirectory >>= canonicalizePath
+      getRelativeOrAbsoluteChild current path @?= "dir1/dir2",
+    testCase "Mixed indirections result in current directory" $ do
+      path <- canonicalizePath "././dir1/./.././dir2/./../"
+      current <- getCurrentDirectory >>= canonicalizePath
+      getRelativeOrAbsoluteChild current path @?= "./",
+    testCase "Child directory" $ do
+      path <- canonicalizePath "./dir1/dir2/"
+      current <- getCurrentDirectory >>= canonicalizePath
+      getRelativeChild current path @?= Just "dir1/dir2",
+    testCase "Not a child directory" $ do
+      root <- canonicalizePath "./dir1/dir2/"
+      path <- canonicalizePath "./dir1/dir3/"
+      getRelativeChild root path @?= Nothing,
+    testCase "No intermediate directories same" $ do
+      path <- canonicalizePath "./dir1/dir3/"
+      getDirsBetweenRootAndFile path path @?= [],
+    testCase "No intermediate directories different" $ do
+      root <- canonicalizePath "./dir1/dir2/"
+      path <- canonicalizePath "./dir1/dir3/"
+      getDirsBetweenRootAndFile root path @?= [],
+    testCase "Intermediate directories" $ do
+      path <- canonicalizePath "./example/dir/other"
+      current <- getCurrentDirectory >>= canonicalizePath
+      getRelativeOrAbsoluteChild current <$> getDirsBetweenRootAndFile current path @?=
+        [ "./"
+        , "example"
+        , "example/dir"
+        ]
+  ]
```

Comment:
Some minor suggestion: `test_*` function can also be of `IO TestTree` type, you could extract `getCurrentDirectory >>= canonicalizePath` there.

Also, would be nice to group test cases by what function you test (if any particular) using `testGroup`.